File: binV2/fourth_problem.rb
require 'json'

input_data = [
  { "name": "john doe", "email": "john@example.com" },
  { "name": "John Doe", "email": "JOHN@example.com" },
  { "name": "jane smith", "email": "jane@example.com" }
]

grouped_data = input_data.group_by { |data| [ data[:name].downcase, data[:email].downcase ] }

duplicated_data = grouped_data.select { |k, v| v.size > 1 }.values.flatten

print JSON.pretty_generate({ data: duplicated_data })

File: quizzes4/fourth_problem.js
const customers = [
  { name: "john doe", email: "john@example.com" },
  { name: "John Doe", email: "JOHN@example.com" },
  { name: "jane smith", email: "jane@example.com" },
];

const findDuplicatesWithObj = () => {
  const groupedCustomers = Object.groupBy(customers, (cust) => [
    cust?.name.toLowerCase(),
    cust?.email.toLowerCase(),
  ]);

  return Object.values(groupedCustomers).flatMap((cust) =>
    cust.length > 1 ? cust : [],
  );
};

const findDuplicatesWithLoop = () => {
  const seen = new Map();
  const duplicateCustomers = [];

  for (let i = 0; i < customers.length; i++) {
    const cust = customers[i];
    const key = `${cust?.name.toLowerCase()}|${cust?.email.toLowerCase()}`;

    if (!seen.has(key)) {
      seen.set(key, { data: cust, addedToDup: false });
    } else {
      const entry = seen.get(key);

      if (!entry.addedToDup) {
        duplicateCustomers.push(entry.data);
        entry.addedToDup = true;
      }

      duplicateCustomers.push(cust);
    }
  }

  return duplicateCustomers;
};

console.log(JSON.stringify({ data: findDuplicatesWithObj() }, null, 2));
console.log(JSON.stringify({ data: findDuplicatesWithLoop() }, null, 2));
